### PR TITLE
Unit Tests for sync, range sync and file close failures

### PR DESCRIPTION
**Pull request by siying**

### CODE REVIEW

File: db/db_io_failure_test.cc [hunk 1]
```diff
@@ -253,6 +253,203 @@ TEST_F(DBIOFailureTest, PutFailsParanoid) {
...
+  ASSERT_OK(Put(1, "foo1_3", rnd_str));
+  ASSERT_OK(Put(1, "foo2", "bar"));
+  ASSERT_OK(Put(1, "foo3_1", rnd_str));
+  ASSERT_OK(Put(1, "foo3_2", rnd_str));
+  ASSERT_OK(Put(1, "foo3_3", rnd_str));
+  ASSERT_OK(Put(1, "foo4", "bar"));
+  dbfull()->TEST_WaitForFlushMemTable(handles_[1]);
+
+  // Following writes should fail as flush failed.
+  ASSERT_NOK(Put(1, "foo2", "bar3"));
```
Comment (IslamAbdelRahman) [inline db/db_io_failure_test.cc]: Is it possible that this test will be flaky ?

**Comment by ajkr** [comment]
I'll review on Monday. Just wanted to mention, when repro'ing the issue, I most often see EIO in the fdatasync() just before close(). I saw it fail the SST file creation many times, so it doesn't matter much to me whether we unit test that one or not.

File: db/db_io_failure_test.cc (hunk 1 above)
Comment (siying) [inline db/db_io_failure_test.cc]: Why?

**Comment by siying** [comment]
@ajkr OK. Should add that one too.

File: db/db_test_util.h [hunk 2]
```diff
@@ -247,6 +247,13 @@ class SpecialEnv : public EnvWrapper {
         }
       }
       Status Truncate(uint64_t size) override { return base_->Truncate(size); }
+      Status RangeSync(uint64_t offset, uint64_t nbytes) override {
+        Status s = base_->RangeSync(offset, nbytes);
+#if !(defined NDEBUG) || !defined(OS_WIN)
+        TEST_SYNC_POINT_CALLBACK("SpecialEnv::SStableFile::RangeSync", &s);
```
Comment (ajkr) [inline db/db_test_util.h]: reminds me of our future api for overriding bg_error_ :).

File: db/db_io_failure_test.cc [hunk 3]
```diff
@@ -253,6 +253,203 @@ TEST_F(DBIOFailureTest, PutFailsParanoid) {
...
+      });
+  rocksdb::SyncPoint::GetInstance()->EnableProcessing();
+
+  ASSERT_OK(dbfull()->SetOptions(handles_[1],
+                                 {
+                                     {"disable_auto_compactions", "false"},
+                                 }));
+  dbfull()->TEST_WaitForCompact();
+
+  // Following writes should fail as flush failed.
```
Comment (ajkr) [inline db/db_io_failure_test.cc]: s/flush/compaction/

File: db/db_io_failure_test.cc [hunk 4]
```diff
@@ -253,6 +253,203 @@ TEST_F(DBIOFailureTest, PutFailsParanoid) {
...
+  ASSERT_OK(Put(1, "foo", "bar"));
+  ASSERT_OK(Put(1, "foo2", "bar"));
+  Flush(1);
+  ASSERT_OK(Put(1, "foo", "bar2"));
+  ASSERT_OK(Put(1, "foo2", "bar"));
+  Flush(1);
+  ASSERT_OK(Put(1, "foo", "bar3"));
+  ASSERT_OK(Put(1, "foo2", "bar"));
+  Flush(1);
+  dbfull()->TEST_WaitForCompact();
```
Comment (ajkr) [inline db/db_io_failure_test.cc]: delete

**Comment by siying** [comment]
@ajkr you are right. Maybe I should consolidate the two places.

**Pull request closed without merging**